tests/unittests/algo/test_evolution_es.py

```diff
@@ -237,3 +182,83 @@ def test_register(self, evolution, bracket, rung_0, rung_1):
         assert len(bracket.rungs[0])
         assert point_hash in bracket.rungs[0]['results']
         assert (0.0, point) == bracket.rungs[0]['results'][point_hash]
+
+
+class TestBracketEVES():
+    """Tests for `BracketEVES` class.."""
+
+    def test_get_teams(self, bracket, rung_3):
+        """Test that correct team is promoted."""
+        bracket.rungs[0] = rung_3
+        rung, population_range, red_team, blue_team = bracket._get_teams(0)
+        assert len(list(rung.values())) == 4
+        assert bracket.search_space_remove_fidelity == [1, 2]
+        assert population_range == 4
+        assert set(red_team).union(set(blue_team)) == {0, 1, 2, 3}
+        assert set(red_team).intersection(set(blue_team)) == set()
+
+    def test_get_mutated_population(self, bracket, rung_3):
+        """Verify mutated candidates is generated correctly."""
+        red_team = [0, 2]
+        blue_team = [1, 3]
+        for i in range(4):
+            for j in [1, 2]:
+                bracket.eves.population[j][i] = list(rung_3["results"].values())[i][1][j]
+            bracket.eves.performance[i] = list(rung_3["results"].values())[i][0]
+
+        org_data = np.stack((list(bracket.eves.population.values())[0],
+                             list(bracket.eves.population.values())[1]), axis=0).T
+
+        org_data = copy.deepcopy(org_data)
+
+        bracket._get_mutated_population(red_team, blue_team)
+
+        mutated_data = np.stack((list(bracket.eves.population.values())[0],
+                                 list(bracket.eves.population.values())[1]), axis=0).T
+
+        assert org_data.shape == mutated_data.shape
+        assert (mutated_data[0] == org_data[0]).all()
+        assert (mutated_data[2] == org_data[2]).all()
+        assert (mutated_data[1] != org_data[1]).any()
+        assert (mutated_data[3] != org_data[3]).any()
+        assert (mutated_data[1] != org_data[0]).any()
+        assert (mutated_data[3] != org_data[2]).any()
+
+        if mutated_data[1][0] != org_data[0][0]:
+            assert mutated_data[1][1] == org_data[0][1]
+        else:
+            assert mutated_data[1][1] != org_data[2][1]
+
+        if mutated_data[3][0] != org_data[2][0]:
+            assert mutated_data[3][1] == org_data[2][1]
+        else:
+            assert mutated_data[3][1] != org_data[2][1]
+
+    def test_duplicated_mutated_population(self, bracket, rung_4):
+        """Verify duplicated candidates can be found and processed correctly."""
+        population_range = 4
+        for i in range(4):
+            for j in [1, 2]:
+                bracket.eves.population[j][i] = list(rung_4["results"].values())[i][1][j]
+        points, nums_all_equal = bracket._get_mutated_points(rung_4["results"], population_range)
+
+        if points[1][1] != points[2][1]:
```

Comment:
Could you please document the rational here for the tests as well. Thanks!

Reply:
Done. In this case, duplication will occur, and we can make it mutate one more time. The points 1 and 2 should be different, while one of nums_all_equal should be 1 (meaning it mutate one more time). I added this comments in code.